### ISSUE TITLE
problems with vue 2.0 and `Vue.set` which now requires 3 parameters

### ISSUE THREAD

**Issue by admataz** [revue/revue#19]
Hi, 
I've been experimenting with using Revue with the most recent RC for Vue 2.0 and have hit the following problem.

Revue sets the value on a Vue component here: 
https://github.com/revue/revue/blob/v2.2.0/src/revue.js#L35

``` javascript
this.$set(realProp, deepProp(store.getState(), storeProp))
```

The data object on my Vue component is defined like this:

``` javascript
...
data() {
    return {
        message: this.$select('message')
    }
  }

```

Which causes an error when an action is dispatched by the store, as described here: 

https://github.com/vuejs/vue/pull/3659

Something like: 
`vue.js:1126 Uncaught TypeError: Cannot create property 'here is a test messasge' on string 'message'`

In my tinkering I have found updates to the component work if I define and target a property on the component's  `data` object for the Redux store and passing this as the object in the above line 

Something like: 

``` javascript
_this.$set(_this.store, realProp, deepProp(store.getState(), storeProp));
```

and in my component definiton, the data object is defined like this:: 

``` javascript
...
data() {
    return {
      store: {
        message: this.$select('message')
      }
    }
  }

```

This seems a limiting/hacky way to do this - I'm not sure if there's a better way, or if Vue 2.0 supports updating properties on the 'data'  object?

But as it stands, I can only use this library with Vue 1

**Comment by aweber1** [comment]
I encountered this same issue and fixed it by changing the line mentioned above (https://github.com/revue/revue/blob/v2.2.0/src/revue.js#L35) from

`this.$set(realProp, deepProp(store.getState(), storeProp))` 

to

`_this[realProp] = deepProp(store.getState(), storeProp);`

And all is well.

Also, the readme seems to be inaccurate around how to select nested state objects as data:
```
// if prop is not in top level
// do this.$select('todos as path.to.todos')
```

should be:

```
// if prop is not in top level
// do this.$select('path.to.todos as todos')
```

I'd be happy to submit a PR for these changes, but wasn't sure if @egoist wanted to open up a new branch for Vue 2.0-specific changes or not.

**Comment by egoist** [comment]
@aweber1 I'd like to make you an collaborator if you are interested in working on 2.0 branch.

**Comment by admataz** [comment]
@aweber1  - do you know if bypassing Vue's `$set` method with this direct assignment of an object property is safe in terms of Vue's reactivity model? 

I am under the impression that direct assignment could cause issues (see [Reactivity in Depth - vue.js - Change Detection Caveats](http://rc.vuejs.org/guide/reactivity.html#Change-Detection-Caveats))

If my understanding is correct, the direct replacement method would replace the Vue instance's root `data` property with a new object (as redux store returns a copy)  - which has not been initialised with Vue's getter/setter conversion.

**Comment by aweber1** [comment]
@admataz I had the exact same question and tried this: 
```
_this.$set(_this._data, realProp, deepProp(store.getState(), storeProp))
``` 
but Vue complained about attempting to dynamically add a property to the component data object.

So I dug in further to the v2.0 docs (the same 'Reactivity in Depth' section you linked to http://rc.vuejs.org/guide/reactivity.html#Change-Detection-Caveats and the 'Declaring Reactive Properties' section http://rc.vuejs.org/guide/reactivity.html#Declaring-Reactive-Properties) and it seems to indicate that you just need to declare your data/reactive properties upfront and then anytime later you can directly set/modify the value of those properties.

I could be completely wrong and would happily accept that :) I'm by no means a Vue expert, fumbling my way through a Vue 2.0 PoC at the moment.

**Comment by admataz** [comment]
@aweber1  - looking deeper into this myself  - I found setting a property value directly the way you have _is_ the supported/preferred method by Vue to update values for _existing_ properties on an instance (as long as they have been declared in the data object at instantiation). Vue establishes getters and setters for the properties to make them reactive.   It is the only way to update a top level property. 

The complication comes with adding nested properties - which a new Redux state could possibly do.

`Vue.set` (or `vm.$set`) is a method to get around limitations when adding new _nested_ properties and dealing with mutations to arrays. Mutations to arrays are not an issue to Redux - but adding new nested properties and making them reactive is a potential issue. 

My conclusion:
- for setting nested properties we should use `Vue.set` (or `this.$set`)  in case there are new nested properties: 

`this.$set(this.path.to.object, propName, valueFromStore);`

-  for top-level properties, we should assign the new value as you have in your example: 
 
`this[topLevelProperty] = valueFromStore`

I hope that makes sense - and that I've understood Vue correctly.

**Comment by villeristi** [comment]
@egoist, @aweber1, @admataz what is the conclusion on this one? Is there going to be modifications for Vue 2, or how should one use this library?

**Comment by pajtai** [comment]
can confirm that @caspersleep 's fork works, which has the modification described above

**Comment by egoist** [comment]
should be fixed in `v3.0.0`
